// @remotion/skia/dist/cjs/enable.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.enableSkia = void 0;
const bundler_1 = require("@remotion/bundler");
const fs_1 = __importDefault(require("fs"));
/**
 * @description A function that modifies the default Webpack configuration to make the necessary changes to support Skia.
 * @see [Documentation](https://www.remotion.dev/docs/skia/enable-skia)
 */
const enableSkia = (currentConfiguration) => {
    var _a, _b, _c, _d;
    const newExtensions = [
        '.web.js',
        '.web.ts',
        '.web.tsx',
        ...((_b = (_a = currentConfiguration.resolve) === null || _a === void 0 ? void 0 : _a.extensions) !== null && _b !== void 0 ? _b : []),
    ];
    const deduplicatedExtensions = [...new Set(newExtensions)];
    return {
        ...currentConfiguration,
        plugins: [
            ...((_c = currentConfiguration.plugins) !== null && _c !== void 0 ? _c : []),
            new (class CopySkiaPlugin {
                apply(compiler) {
                    compiler.hooks.thisCompilation.tap('AddSkiaPlugin', (compilation) => {
                        compilation.hooks.processAssets.tapPromise({
                            name: 'copy-skia',
                            stage: compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL,
                        }, async () => {
                            const src = require.resolve('canvaskit-wasm/bin/full/canvaskit.wasm');
                            if (compilation.getAsset(src)) {
                                // Skip emitting the asset again because it's immutable
                                return;
                            }
                            compilation.emitAsset('/canvaskit.wasm', new bundler_1.webpack.sources.RawSource(await fs_1.default.promises.readFile(src)));
                        });
                    });
                }
            })(),
        ],
        resolve: {
            ...currentConfiguration.resolve,
            // FIXME: To fix missing modules in browser when using webassembly
            fallback: {
                fs: false,
                path: false,
            },
            extensions: deduplicatedExtensions,
            alias: {
                ...(_d = currentConfiguration.resolve) === null || _d === void 0 ? void 0 : _d.alias,
                'react-native-reanimated': "require('react-native-reanimated')",
                'react-native-reanimated/lib/reanimated2/core': "require('react-native-reanimated/lib/reanimated2/core')",
                'react-native/Libraries/Image/AssetRegistry': false,
            },
        },
    };
};
exports.enableSkia = enableSkia;
